### PR TITLE
Feature/upscale

### PR DESCRIPTION
**Pull request by namoray**
Added new upscale workflow, doesnt require prompt. Adds details with sampling. (New bool parameter with upscale endpoint)
Fixed avatar issues with same background, stretching and weird results sometimes.
Added complete caching to avatar if using highvram.
Fixed img2img dreamshaper workflow.
Saving input images as unique id and deleting upon inference.

### CODE REVIEW

File: Dockerfile.image_server.with_models [hunk 1]
```diff
@@ -67,8 +67,10 @@ ENV HF_HUB_ENABLE_HF_TRANSFER=1
 USER root
 ENV VLLM_PORT=6919
 ENV DEVICE=0
+ENV VRAM_MODE=--highvram
```
Comment (namoray) [inline Dockerfile.image_server.with_models:70]: This is up to users, so the default doesn't belong here

File: Dockerfile.image_server.with_models [hunk 2]
```diff
@@ -67,8 +67,10 @@ ENV HF_HUB_ENABLE_HF_TRANSFER=1
...
+ENV WARMUP=true
```
Comment (namoray) [inline Dockerfile.image_server.with_models:71]: Same with this

File: Dockerfile.image_server.with_models [hunk 3]
```diff
@@ -67,8 +67,10 @@ ENV HF_HUB_ENABLE_HF_TRANSFER=1
...
 
 RUN chmod +x /app/image_server/entrypoint.sh
 
 
-CMD ["bash", "-c", "./app/image_server/entrypoint.sh"]
```
Comment (namoray) [inline Dockerfile.image_server.with_models:74]: Why is this changed - does it fail without it?

File: image_server/entrypoint.sh [hunk 4]
```diff
@@ -36,11 +36,16 @@ sleep 5
...
+    else
+        python warmup.py
+    fi
 else
     sleep 1
 fi
 
 cd /app/image_server/
-uvicorn main:app --host 0.0.0.0 --port $port
+uvicorn main:app --host 0.0.0.0 --port $port --workers 1
```
Comment (namoray) [inline image_server/entrypoint.sh:50]: Why?

File: image_server/entrypoint.sh [hunk 5]
```diff
@@ -36,11 +36,16 @@ sleep 5
 if [ "$warmup" = "true" ]
 then
     cd /app/image_server/
-    python warmup.py
+    if [ -n "$vram_mode" ] && [ "$vram_mode" = "--highvram" ]
```
Comment (namoray) [inline image_server/entrypoint.sh:39]: Surely we want to do this for --gpu-only too? Also why do you need both check its valid and check its highvram?

File: image_server/payload.py [hunk 6]
```diff
@@ -128,23 +138,33 @@ def modify_img2img(self, input_data: Img2ImgBase) -> Dict[str, Any]:
...
+            init_img = base64_to_image(input_data.init_image)
+            img_id = uuid.uuid4()
+            init_img.save(f"{cst.COMFY_INPUT_PATH}{img_id}.png")
+            payload["Image_loader"]["inputs"]["image"] = f"{img_id}.png"
+        except Exception as e:
+            traceback.print_exc()
+        return payload, [img_id]
+
+    def modify_avatar(self, input_data: AvatarBase) -> Tuple[Dict[str, Any], List[str]]:
+        avatar_workflow = "instantid_highvram" if ((torch.cuda.get_device_properties(cst.DEFAULT_DEVICE).total_memory / 1e9) > 25.0) else "instantid"
```
Comment (namoray) [inline image_server/payload.py:157]: What's this for? Why are we checking the memory of the device rather than seeing what mode we're in?

File: image_server/payload.py [hunk 7]
```diff
@@ -110,10 +119,11 @@ def modify_txt2img(self, input_data: Txt2ImgBase) -> Dict[str, Any]:
         payload["Latent"]["inputs"]["height"] = input_data.height
         return payload
 
-    def modify_img2img(self, input_data: Img2ImgBase) -> Dict[str, Any]:
+    def modify_img2img(self, input_data: Img2ImgBase) -> Tuple[Dict[str, Any], List[str]]:
         payload = copy.deepcopy(self._payloads[f"img2img_{input_data.engine}"])
         init_img = base64_to_image(input_data.init_image)
-        init_img.save(f"{cst.COMFY_INPUT_PATH}init.png")
+        img_id = uuid.uuid4()
```
Comment (namoray) [inline image_server/payload.py:125]: Rather than having this in every funciton, this should be some generic funciton which will handle this (getting a uuid and saving the image

File: image_server/payload.py [hunk 8]
```diff
@@ -128,23 +138,33 @@ def modify_img2img(self, input_data: Img2ImgBase) -> Dict[str, Any]:
...
-    def modify_avatar(self, input_data: AvatarBase) -> Dict[str, Any]:
-        payload = copy.deepcopy(self._payloads["instantid"])
+        payload["Image_loader"]["inputs"]["image"] = f"{img_id}.png"
+        return payload, [img_id]
+
+    def modify_upscale(self, input_data: UpscaleBase) -> Tuple[Dict[str, Any], List[str]]:
+        try:
+            upscale_mode = "upscale_sampled" if input_data.sampled else "upscale"
+            payload = copy.deepcopy(self._payloads[upscale_mode])
+            init_img = base64_to_image(input_data.init_image)
```
Comment (namoray) [inline image_server/payload.py:148]: Same here, this shouldn't be here but in some generic function if its used in lots of places

File: image_server/warmup.py [hunk 9]
```diff
@@ -2,19 +2,33 @@
...
-    if filename.endswith(".json"):
-        filepath = os.path.join(directory, filename)
-        with open(filepath, "r") as file:
-            try:
-                payload = json.load(file)
-                image = api_gate.generate(payload)
-            except json.JSONDecodeError as e:
-                print(f"Error decoding JSON from {filename}: {e}")
+
+if args.highvram:
```
Comment (namoray) [inline image_server/warmup.py:16]: What about gpu only? WHy do we need a different workflow warmup for highvram mode?

File: Dockerfile.image_server.with_models (hunk 1 above)
Comment (diagonalge) [reply Dockerfile.image_server.with_models:70]: noted, will remove

File: Dockerfile.image_server.with_models (hunk 2 above)
Comment (diagonalge) [reply Dockerfile.image_server.with_models:71]: noted

File: Dockerfile.image_server.with_models (hunk 3 above)
Comment (diagonalge) [reply Dockerfile.image_server.with_models:74]: nothing really changed here

File: image_server/entrypoint.sh (hunk 4 above)
Comment (diagonalge) [reply image_server/entrypoint.sh:50]: Making sure that running 1 worker ensures that we are processing one request at a time and not sending then asynchronously to ComfyUI

File: image_server/entrypoint.sh (hunk 5 above)
Comment (diagonalge) [reply image_server/entrypoint.sh:39]: This workflow is explicitly meant to warmup on higher end gpus, can do this for --gpu-only too yes. This basically loads everything in the VRAM

File: image_server/payload.py (hunk 6 above)
Comment (diagonalge) [reply image_server/payload.py:157]: Yeah this was up for discussion, highvram mode is limited to ComfyUI's domain. I think we can add a global to fastAPI for this stuff. What do you think?

File: image_server/payload.py (hunk 7 above)
Comment (diagonalge) [reply image_server/payload.py:125]: noted

File: image_server/payload.py (hunk 8 above)
Comment (diagonalge) [reply image_server/payload.py:148]: what exactly? selecting the payload?

File: image_server/warmup.py (hunk 9 above)
Comment (diagonalge) [reply image_server/warmup.py:16]: yes can do gpu-only instead of highvram. This is a warmup that caches basically everything and needs high vram. I am caching the controlnet model and ipadapter model for instantid in the avatar workflow with this warmup. The problem is that it MUST run if we want to cache everything because I am retrieving the cache in the avatar workflow.

File: image_server/entrypoint.sh (hunk 4 above)
Comment (namoray) [reply image_server/entrypoint.sh:50]: Does it? I think the uvicorn server already only has 1 worker

File: image_server/entrypoint.sh (hunk 5 above)
Comment (namoray) [reply image_server/entrypoint.sh:39]: higher end gpus != highvram mdoe

high vram mode is just when we want to keep stuff in gpu - that could be on 4090s or a100's.

But still, why do we need a separate warmup, why is it a problem if they're on 4090s with low vram mode?

File: image_server/payload.py (hunk 6 above)
Comment (namoray) [reply image_server/payload.py:157]: I'm not sure what you mean - but device memory has nothing to do (ish) to whether we want to run highvram mode or not - i can run just avatar on a 4090 and want to use --gpuonly for exampke

File: image_server/payload.py (hunk 8 above)
Comment (namoray) [reply image_server/payload.py:148]: creating a uid, saving the image, modifying hte payload based on that. See how we do positive and negative prompts for example which is a generic func

File: image_server/warmup.py (hunk 9 above)
Comment (namoray) [reply image_server/warmup.py:16]: Not instead of , those are both options in comfyui right

What if we don't want to cache everything, can't we run this workflow anyway, and if we're on low vram mode, it wont cache it. Is that not simpler?

File: image_server/entrypoint.sh (hunk 5 above)
Comment (diagonalge) [reply image_server/entrypoint.sh:39]: 24 gbs is low for this warmup. We can just do normal one in that case?

File: image_server/payload.py (hunk 6 above)
Comment (diagonalge) [reply image_server/payload.py:157]: okay, i will add the vram mode to the fastapi to decide all this.

File: image_server/payload.py (hunk 8 above)
Comment (diagonalge) [reply image_server/payload.py:148]: noted

File: image_server/warmup.py (hunk 9 above)
Comment (diagonalge) [reply image_server/warmup.py:16]: Yes sounds like a plan. I will use this warmup always then.

File: image_server/entrypoint.sh (hunk 4 above)
Comment (diagonalge) [reply image_server/entrypoint.sh:50]: Yup just made sure that it doesn't spawn multiple workers that send async requests to ComfyUI. In my testing the requests are being processed one by one.

File: image_server/entrypoint.sh (hunk 5 above)
Comment (namoray) [reply image_server/entrypoint.sh:39]: you need more than 24gb of ram for this model to work?

File: image_server/entrypoint.sh (hunk 4 above)
Comment (namoray) [reply image_server/entrypoint.sh:50]: gotcha

File: image_server/payload.py (hunk 8 above)
Comment (namoray) [reply image_server/payload.py:148]: can you do it then pleasE :D

File: image_server/entrypoint.sh (hunk 5 above)
Comment (diagonalge) [reply image_server/entrypoint.sh:39]: This warmup, yes.. won't throw an error just some things won't be cached
